### ISSUE TITLE
Asset Pipeline: Bundling all js or css files but skip compilation/minification of pre-build files

### ISSUE THREAD

**Issue by pamtbaau** [getgrav/grav#2077]
I'm using Bootstrap v4 and bootstrap's own build scripts that use Sass, Autoprefixer and UglifyJS with their specific settings. This gives more flexibility to alter Bootstrap's defaults (colors, breakpoints, etc.) and/or allows for leaving out unused js/css files.

I'm unsure if Grav's pipeline conflicts with Bootstrap's compilation requirements and therefore would like to add the custom compiled/minified Bootstrap assets without Grav's pipeline, but with the bundling of all files in the end.

I hoped `{% do assets.addCss('theme://css/style.min.css', { priority: 1, pipeline: false, group: 'bottom' }) %}` would bundle style.min.css with other css files targeting 'bottom' but unfortunately it doesn't.

Am I overlooking an option to bundle all css resp. js files but skip Grav's compilation/minifying process on my pre-processed files?

**Comment by rhukster** [comment]
It probably is not working because there's no `{{ assets.css('bottom') }}` in your theme.. This is understandable because CSS generally is only output in one place.

Just do: ` {% do assets.addCss('theme://css/style.min.css', { priority: 1, pipeline: false) %}`

**Comment by pamtbaau** [comment]
I do have a `{{ assets.css('bottom') }}` in my theme...

To recap my aim:
- No processing of pre-compiled `style.min.css`.
It has had its own node-sass/postcss/minify process.
- Bundle preserved `style.min.css` and all "pipelined" files from same `group` into single output file

NB. Same aim goes for javascript files.

Below are some tests. All files have the same group, only pipeline false/true changes.

**TLDR:**
- Only files with `pipeline: true` are bundled inside a given group
If this is as expected, would it be an idea to add an extra option `bundle: boolean` to force bundling of non-pipelined with pipelined files of same group?
- Priority is not preserved between pipelined/non-pipelined files.

**Long read:**
And here are the tests.

**No files pipelined**
```
{% do assets.addCss('theme://css/style.min.css', { priority: 3, pipeline: false, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom1.css',   { priority: 2, pipeline: false, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom2.css',   { priority: 1, pipeline: false, group: 'bottom' }) %}
{{ assets.css('bottom') }}
```
generates:
```
<link href="/user/themes/basetheme/css/style.min.css" type="text/css" rel="stylesheet" />
<link href="/user/themes/basetheme/css/custom1.css" type="text/css" rel="stylesheet" />
<link href="/user/themes/basetheme/css/custom2.css" type="text/css" rel="stylesheet" />
```
Note:
- No bundling at all
- Ordering according priority

---

**One file is pipelined**
```
{% do assets.addCss('theme://css/style.min.css', { priority: 3, pipeline: false, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom1.css',   { priority: 2, pipeline: false, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom2.css',   { priority: 1, pipeline: true, group: 'bottom' }) %}
{{ assets.css('bottom') }}
```
generates:
```
<link href="/assets/f9579cb9d6d0ef60322a5a1e738a7742.css" type="text/css" rel="stylesheet" />
<link href="/user/themes/basetheme/css/style.min.css" type="text/css" rel="stylesheet" />
<link href="/user/themes/basetheme/css/custom1.css" type="text/css" rel="stylesheet" />
```
Note:
- No bundling at all
- Priority of pipelined css file is wrong
---

**Two files pipelined.**
Hoped this would match my aim.
```
{% do assets.addCss('theme://css/style.min.css', { priority: 3, pipeline: false, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom1.css',   { priority: 2, pipeline: true, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom2.css',   { priority: 1, pipeline: true, group: 'bottom' }) %}
{{ assets.css('bottom') }}
```
generates:
```
<link href="/assets/6538007958629adeeff8d044cbe34770.css" type="text/css" rel="stylesheet" />
<link href="/user/themes/basetheme/css/style.min.css" type="text/css" rel="stylesheet" />
```
Note:
- Only bundling of pipelined css files
- Priority of pipelined and bundled css file is wrong
- Priority preserved inside bundled css

---

** All files pipelined.**
```
{% do assets.addCss('theme://css/style.min.css', { priority: 3, pipeline: true, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom1.css',   { priority: 2, pipeline: true, group: 'bottom' }) %}
{% do assets.addCss('theme://css/custom2.css',   { priority: 1, pipeline: true, group: 'bottom' }) %}
{{ assets.css('bottom') }}
```
generates:
```
<link href="/assets/8b868ccb5edd7deefa4345c7a3d8a0be.css" type="text/css" rel="stylesheet" />
```
Note:
- All css files are bundled
- Priority preserved inside bundled css

**Comment by rhukster** [comment]
What you are seeing is actually expected.

Pipeline = compress and bundle together.

If you have `pipeline: false` that will ensure the entry is not included in the pipeline/bundle. 

Order should work for pipeline and non-pipelined assets, but its not smart enough to know how to pipeline some assets, then a non-pipelined asset, then continue with a new piplined file for assets after. 

Its simply outputs non-pipelined stuff, then pipelined stuff.  There is a config option (under `assets:`) that lets you configure:

```
css_pipeline_before_excludes:
js_pipeline_before_excludes:
```

These options flip the order.

**Comment by pamtbaau** [comment]
Question:
- How do I preserve my compilation process **ánd** bundle with other files in the same group?

Request:
- Add option `bundle: boolean` to addCss/addJs to override bundling behavior.

NB. Thanks for pointing me to `css_pipeline_before_excludes` and `js_pipeline_before_excludes`

**Comment by pamtbaau** [comment]
@rhukster Sorry for spending your time on this rookie Grav-ist issue. I have misperceived the activities of `pipeline` and was unsure it might hamper my pre-processed scss files.

Looking at the Assets class and appreciating its benign character cleared that up...
